opencog/query/PatternUtils.cc

```diff
@@ -164,6 +164,20 @@ void get_connected_components(const std::set<Handle> &vars,
 	compset.insert(components.begin(), components.end());
 }
 
+void split_clauses_pos_neg(const std::vector<Handle>& clauses,
+                           std::vector<Handle>& affirm,
+                           std::vector<Handle>& negate)
+{
+	for (Handle h : clauses) {
+		if (NOT_LINK == h->getType()) {
```

Comment:
OK .. so, please realize that the "NotLink" here is a "temporary hack" mostly because I could not think of a better solution.  What is really needed is maybe(?) an "OptionalLink" or something like that .... something that says "if this part of the pattern is found then call optional_clause_match() instead of clause_match()".

The problem with NotLink is philosophical/theoretical: it does not play nice with so-called "intuitionsitic logic" or "constructive logic"  and I'm trying hard to make sure that opencog  defaults to constructive logic instead of classical logic.  The problem with classical logic is that it makes certain types of reasoning (extremely, disaterously) inefficient by blocking the use of certain fast algos. 

Creating an OptionalLink or a MaybeLink is the best I can think of right now, ...